code/espurna/sensors/EZOPHSensor.h

```diff
@@ -0,0 +1,142 @@
+
+    public:
+
+        // ---------------------------------------------------------------------
+        // Public
+        // ---------------------------------------------------------------------
+
+        EZOPHSensor(): BaseSensor() {
+            _count = 1;
+            _sensor_id = SENSOR_EZOPH_ID;
+        }
+
+        ~EZOPHSensor() {
+            if (_serial) delete _serial;
+        }
+
+        // ---------------------------------------------------------------------
+
+        void setRX(unsigned char pin_rx) {
+            if (_pin_rx == pin_rx) return;
+            _pin_rx = pin_rx;
+            _dirty = true;
+        }
+
+        void setTX(unsigned char pin_tx) {
+            if (_pin_tx == pin_tx) return;
+            _pin_tx = pin_tx;
+            _dirty = true;
+        }
+
+        // ---------------------------------------------------------------------
+
+        unsigned char getRX() {
+            return _pin_rx;
+        }
+
+        unsigned char getTX() {
+            return _pin_tx;
+        }
+
+        // ---------------------------------------------------------------------
+        // Sensor API
+        // ---------------------------------------------------------------------
+
+        // Initialization method, must be idempotent
+        void begin() {
+            if (!_dirty) return;
+
+            if (_serial) delete _serial;
+
+            _serial = new SoftwareSerial(_pin_rx, _pin_tx);
+            _serial->begin(9600);
+
+            _ready = true;
+            _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() {
+            char buffer[28];
+            snprintf(buffer, sizeof(buffer), "EZOPH @ SwSerial(%u,%u)", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Descriptive name of the slot # index
+        String slot(unsigned char index) {
+            return description();
+        };
+
+        // Address of the sensor (it could be the GPIO or I2C address)
+        String address(unsigned char index) {
+            char buffer[6];
+            snprintf(buffer, sizeof(buffer), "%u:%u", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {
+            if (index == 0) return MAGNITUDE_PH;
+            return MAGNITUDE_NONE;
+        }
+
+        void pre() {
+            _read();
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            if (index == 0) return _ph;
+            return 0;
+        }
+
+
+
+    protected:
+
+        // ---------------------------------------------------------------------
+        // Protected
+        // ---------------------------------------------------------------------
+
+        void _read() {
+            String ph_serial = "";
+            ph_serial.reserve(30);
+
+            while ((_serial->available() > 0)) {
```

Comment:
The device is in continuous read mode by default. 1 reading per second. Is this the right way to read as fast as possible?

Reply:
#referencing https://github.com/plerup/espsoftwareserial/blob/master/SoftwareSerial.cpp
If I understand this right, actual reading is done by isr handler and you just advancing the buffer with read()

Reply:
Ah, indeed. I need to look a bit more closely but there is a chance that a larger content of the buffer is returned in a single `read()` right?

Reply:
Tnx to google for reminder - 9600 baud =~ 960 chars per second.
Datasheet does not really mention any other things but short ~5-9char data chunks once in a ~second being `va.lue\r` + `*OK\r`
BTW it should also check response code (which is on by default?)

Reply:
In default continuous mode (UART), the output data are always the readings. There is no `*OK\r`, as that is only in response to commands, which this code does not send.

As far as `*ER`, `*OV` and the other error response codes, I'm thinking of testing the first chard and if it starts with `*`, I will simply set `_error = SENSOR_ERROR_OTHER`.

WDTY?

Reply:
One thing I had noticed in similar [CSE7766 sensor](https://github.com/xoseperez/espurna/blob/88a7076ad3b906f8aeee0ae6978332f0e61b4335/code/espurna/sensors/CSE7766Sensor.h#L291-L336) is that it does the check the buffer every available opportunity using `.tick() -> ._read()`, so it waits only `loop()` time + `delay("loopDelay" / LOOP_DELAY_TIME / 10ms default)` instead. And also discards data on known intervals, counting since the last reading.